### ISSUE TITLE
PKP PLN Network Status is confused by URL aliases

### ISSUE THREAD

**Issue by ctgraham** [pkp/pkp-lib#1423]
@ubermichael, see [my question](http://forum.pkp.sfu.ca/t/pkp-pln-status-question/10932/15?u=ctgraham) and [@mjordan's reply](http://forum.pkp.sfu.ca/t/pkp-pln-status-question/10932/16?u=ctgraham) regarding a journal which has multiple legitimate DNS names.

A problem seems to arise in the use of `$journal->getUrl()` instead of `$base_url`.

**Comment by ctgraham** [comment]
If I'm understanding the concern correctly, we did something similar with usageEvents:
https://github.com/ulsdevteam/ojs/blob/uls-install/plugins/generic/usageEvent/UsageEventPlugin.inc.php#L289-L302

Perhaps this kind of "canonicalUrl" disambiguation should be promoted to the `PKPRequest` object?

**Comment by asmecher** [comment]
Grepping through for uses of `$journal->getUrl`, pretty much all of those use cases should be replaced with the more consistent and dependable `Request::url`. Once that's done, I'd suggest deprecating or removing `$journal->getUrl` entirely.

**Comment by ctgraham** [comment]
Replacing usages of $journal->getUrl() with Request::url() doesn't resolve the disambiguation of multiple base_urls, does it?

**Comment by asmecher** [comment]
Using `Request::url` should cause OJS to forcibly generate URLs per `base_url[...]` overrides, regardless of where the user is coming from.

**Comment by ctgraham** [comment]
Something is not quite right there.  In dev-2_4, `$journal->getUrl()` is just `return Request::url($this->getPath());`  So, if it is nomalizing on base_url or base_url[], the assumption of the issue as reported above is wrong.

**Comment by ctgraham** [comment]
Found it.  Request::url() honors the base_url[] overrides, but ignores the base_url.  In my case, I have just a single journal install, and only base_url is set, no base_url[] overrides.  Apache, however, will direct 5 different DNS entries to this install.

**Comment by asmecher** [comment]
When `base_url[...]` isn't supplied (that's different than `base_url` without `[...]`), OJS will respond according to whatever incoming URL it's invoked from. If you supply `base_url[...]` directives, then that'll force OJS to use whatever is configured as canonical.

**Comment by ctgraham** [comment]
That makes sense to me.

I don't think, however, that as a single journal install I should need to enter a base_url[] override for the PKP PLN to work.  :smile:

**Comment by ubermichael** [comment]
I think the patch linked above gets us most of the way there, although I still haven't been able to replicate @ctgraham's specific problem.

**Comment by ubermichael** [comment]
I think I'm a little out of the loop here. What's up with this issue & PR?

**Comment by ctgraham** [comment]
As of Tuesday of this week, I changed the way our crons run to potentially work around this issue.  Previously, the crons were run with absolute paths:
`/usr/local/bin/php /usr/local/journals/my.journal.name/htdocs/ojs/tools/runScheduledTasks.php /usr/local/journals/my.journal.name/htdocs/ojs/plugins/generic/pln/xml/scheduledTasks.xml`
This caused invalid base_urls to appear in the runs ("/usr/local/journals/my.journal.name/htdocs/" would appear in the "URL")

I still think this is a bug, but now I am running cron from the OJS directory:
`cd /usr/local/journals/my.journal.name/htdocs/ojs/; /usr/local/bin/php tools/runScheduledTasks.php plugins/generic/pln/xml/scheduledTasks.xml`
I have not had a chance to dive into this, but I suspect it will work around the problem.

I'll email you separately to followup on our deposits specifically.

**Comment by asmecher** [comment]
@ubermichael and @ctgraham, can you summarize the state of this one? I see work from both of you (and a PR from Clinton); Michael, I see some commits but no PR.

**Comment by ctgraham** [comment]
So, [PR 1426](https://github.com/pkp/pkp-lib/issues/1427) resolves the problems I was seeing earlier, though I confess I don't understand why it does.  I was afraid the change by itself to propagate "localhost" into the URLs, but it doesn't.

**Comment by ubermichael** [comment]
c30502 removes all uses of `$journal->getUrl()` and replaces them with `$router->url()`. This seems to do the right things in my test setup, when I execute `php tools/runScheduledTasks.php`

I haven't created a PR yet, because I don't know enough about the router to be sure that it works as expected.

**Comment by asmecher** [comment]
Gotcha, @ubermichael. Are you still working on this?

**Comment by asmecher** [comment]
Hmm, I don't think any of us are particularly confident about this one. Is it crucial for OJS 2.4.8-1? (I see it's currently filed against 2.4.9 anyway.)

**Comment by ubermichael** [comment]
The initial problem came about because some (all?) of @ctgraham 's journals were failing to create deposits. It's possible the problem affects anyone else using command-line scripts for cron jobs as well.

**Comment by asmecher** [comment]
OK, that does sound worth fixing, but based on the uncertainty around the fixes here I'm against including it in 2.4.8-1.
